src/coreclr/jit/switchrecognition.cpp

```diff
@@ -132,6 +139,204 @@ bool IsConstantTestCondBlock(const BasicBlock* block,
+        }
+    }
+    return false;
+}
+
+//------------------------------------------------------------------------------
+// optExtendSwitch: If one of the switch's targets is a value test block,
+//    and the switch's variable is the same as the value test's variable - try to merge
+//    that test (BBJ_COND) into the switch block. It's useful to assist Roslyn, which
+//    uses fairly conservative heuristics to generate switch blocks and may give up on
+//    some tests that could be merged into the switch.
+//
+// Arguments:
+//    block - BBJ_SWITCH block to extend
+//
+// Return Value:
+//    True if switch was successfully extended, false otherwise
+//
+bool Compiler::optExtendSwitch(BasicBlock* block)
+{
+    assert(block->KindIs(BBJ_SWITCH));
+    JITDUMP("Considering expanding switch " FMT_BB "\n", block->bbNum);
+
+    GenTree* switchOp           = nullptr;
+    ssize_t  switchTargetOffset = 0;
+    if (!GetSwitchValueOp(block, &switchOp, &switchTargetOffset))
+    {
+        JITDUMP("Couldn't extract the switch's variable - bail out.\n");
+        return false;
+    }
+
+    // TODO-JumpTable: Currently, we only expand the default target, but it should be possible
+    // to traverse all non-default edges as well (if they point to similar value tests).
+    // It shouldn't be too hard to implement (might be TP costly, though).
+
+    BBswtDesc* switchTargets = block->GetSwitchTargets();
+    if (!switchTargets->bbsHasDefault)
+    {
+        JITDUMP("Switch doesn't have a default target - bail out.\n");
+        return false;
+    }
+
+    // Now let's see if we can merge the value test in the edgeToExpand's destination block
+    // with the switch block.
+    FlowEdge*   edgeToExpand    = switchTargets->getDefault();
+    BasicBlock* edgeToExpandDst = edgeToExpand->getDestinationBlock();
+
+    // IsConstantTestCondBlock will tell us if the default target is a value test block (e.g. "x == cns")
+    // TODO-JumpTable: Add support for more complex tests like "x > 10 && x < 20"
+    // Also, we should be able to merge two BBJ_SWITCH on the same variable. Roslyn may create such shapes.
+    FlowEdge* testPasses;
+    FlowEdge* testFails;
+    bool      isReversed;
+    GenTree*  variableNode;
+    ssize_t   cns;
+    if (!IsConstantTestCondBlock(edgeToExpandDst, false, &testPasses, &testFails, &isReversed, &variableNode, &cns))
+    {
+        JITDUMP("The default target is not performing a value test - bail out.\n");
+        return false;
+    }
+
+    BasicBlock* testPassesBb = testPasses->getDestinationBlock();
+    BasicBlock* testFailsBb  = testFails->getDestinationBlock();
+
+    if (!GenTree::Compare(variableNode, switchOp))
+    {
+        JITDUMP("The default target is performing a value test on a different variable - bail out.\n");
+        return false;
+    }
+
+    // We're less conservative than Roslyn, but we still have some limits
+    if (static_cast<size_t>(cns + switchTargetOffset) > SWITCH_MAX_DISTANCE)
```

Comment:
Should we have a stress mode without a limit? I.e., is this correctness or profitability?

Reply:
Yeah, good idea to enable it under stress, it's purely a heuristic

Reply:
Addressed